File: FunctionalPatterns/Structural/Facade.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Structural

module Facade =

    type Cpu = 
        { freeze: unit -> unit
          jump: int64 -> unit
          execute: unit -> unit }
    let newCpu =
        { freeze = fun () -> printfn "Freeze Cpu"
          jump = fun pos -> printfn "Jump %d" pos
          execute = fun () -> printfn "Execute"}

    type HardDrive = { read: int64 -> int32 -> unit }
    let newHardDrive =
        { read = fun lba size -> printfn "Read at %d of size %d" lba size}

    type Memory = { load: int64 -> unit }
    let newMemory =
        { load = fun pos -> printfn "Loading at %d" pos}

    type Computer =
        { start: unit -> unit }
    let newComputer cpu hardDrive memory =
        { start = fun () ->
            let bootSector = 7
            let sectorSize = 42
            let bootAddress = 1234
            cpu.freeze ()
            hardDrive.read bootSector sectorSize
            memory.load bootAddress
            cpu.jump bootAddress
            cpu.execute () }

File: ObjectOrientedPatterns/Structural/Facade.cs
namespace ObjectOrientedPatterns.Structural.Facade
{
    public class Cpu
    {
        public void Freeze() => Console.WriteLine("Freeze CPU");
        public void Jump(long pos) => Console.WriteLine("Jump " + pos);
        public void Execute() => Console.WriteLine("Execute");
    }

    public class HardDrive
    {
        public void Read(long lba, int size) => Console.WriteLine("Read at " + lba + " of size " + size);
    }

    public class Memory
    {
        public void Load(long pos) => Console.WriteLine("Loading at " + pos);
    }

    public class ComputerFacade
    {
        public ComputerFacade(Cpu cpu, HardDrive hardDrive, Memory memory)
        {
            myCpu = cpu;
            myHardDrive = hardDrive;
            myMemory = memory;
        }
        public void Start()
        {
            var bootSector = 7;
            var sectorSize = 42;
            var bootAddress = 1234;
            myCpu.Freeze();
            myHardDrive.Read(bootSector, sectorSize);
            myMemory.Load(bootAddress);
            myCpu.Jump(bootAddress);
            myCpu.Execute();
        }
        private readonly Cpu myCpu;
        private readonly HardDrive myHardDrive;
        private readonly Memory myMemory;
    }
}
